### PR TITLE
added d2m support for reshape and tests

### PR DESCRIPTION
**Pull request by dlokeTT**
### Problem description
Need to support reshape in d2m 

### What's changed
Added support for changing the rank of tensors in `projectLogicalMapToUnitDeviceSpace` which is needed for reshape and `reshapeLogicalMap` which is the affine map for reshape transformations. Assumes minimum rank is 2 for now.
Added builder tests for tile aligned tensors, random shaped tensors and tensors found in llama 3.2 3b. Lots of these are skipped for various reasons, such as compile time #6375, 1d support #6376, and noc unaligned reads/writes #6377.

Had an initial hacky solution for 1d tensors by adding a leading 1 to the metal layout attribute so it would pass the assertion but didn't get to handling all the edge cases so leaving that aside and marking the 1d llama tests as skip for now. I could pick this up when I'm back or we could try and support this natively in the pipeline.

### Checklist
- [ ] New/Existing tests provide coverage for changes

### CODE REVIEW

File: lib/Conversion/TTIRToD2M/TTIRToD2M.cpp [hunk 1]
```diff
@@ -1279,77 +1435,19 @@ class D2MTensorManipulationOpRewriter
...
-    auto newOutTy = RankedTensorType::get(outTy.getShape(),
-                                          outTy.getElementType(), newLayout);
-
-    auto storage =
-        rewriter.create<d2m::EmptyOp>(op.getLoc(), outputs[0].getType());
-    auto view = rewriter.create<d2m::StreamLayoutOp>(
-        op.getLoc(), newOutTy, inputs[0], storage.getResult());
-
-    rewriter.replaceOp(op, unLayoutResult(rewriter, view->getResult(0),
-                                          op->getResult(0).getType()));
```
Comment (nsmithtt) [inline lib/Conversion/TTIRToD2M/TTIRToD2M.cpp:1309]: What was the purpose of this refactor? It's hard to tell if there were any changes made?

File: lib/Conversion/TTIRToD2M/TTIRToD2M.cpp [hunk 2]
```diff
@@ -1279,77 +1435,19 @@ class D2MTensorManipulationOpRewriter
...
-  }
 };
 } // namespace
 
-static AffineMap rearrangeLogicalMap(ttir::RearrangeOp op) {
-  mlir::FailureOr<AffineMap> maybeMap = op.getInvPatternMap();
-  assert(succeeded(maybeMap));
-  return *maybeMap;
-}
-
```
Comment (nsmithtt) [inline lib/Conversion/TTIRToD2M/TTIRToD2M.cpp:1352]: Why move the code?

File: lib/Conversion/TTIRToD2M/TTIRToD2M.cpp (hunk 1 above)
Comment (dlokeTT) [reply lib/Conversion/TTIRToD2M/TTIRToD2M.cpp:1309]: There were no code changes. Was playing around with this and forgot to clean it up when I rebased. Reverted back to how its done in main

File: lib/Conversion/TTIRToD2M/TTIRToD2M.cpp (hunk 2 above)
Comment (dlokeTT) [reply lib/Conversion/TTIRToD2M/TTIRToD2M.cpp:1352]: Also an artifact of some stuff I was playing with and forgot to undo. Put it back to where it was.

File: lib/Conversion/TTIRToD2M/TTIRToD2M.cpp [hunk 3]
```diff
@@ -1313,33 +1313,41 @@ class D2MTensorManipulationOpRewriter
...
+    for (unsigned i = 0; i < inputLogicalRank; ++i) {
+      unsigned revIdx = inputLogicalRank - 1 - i;
+      if (revIdx < 2 && revIdx < outputLogicalRank) {
+        deviceExprs.push_back(
+            builder.getAffineDimExpr(outputLogicalRank - 1 - revIdx));
+      } else if (revIdx < outputLogicalRank) {
+        deviceExprs.push_back(builder.getAffineDimExpr(revIdx));
+      } else {
+        deviceExprs.push_back(builder.getAffineConstantExpr(0));
+      }
```
Comment (nsmithtt) [inline lib/Conversion/TTIRToD2M/TTIRToD2M.cpp:1350]: Is this logically doing what it was doing before?

Comment (dlokeTT) [reply lib/Conversion/TTIRToD2M/TTIRToD2M.cpp:1350]: Oh wait good catch. Ran into issues with this where the grid dim indices would collide for the cases where the ranks of the input and output are the same or the input grid rank is > than the output grid rank. Added handling for each case now where input rank is <, =, > the output rank. This should now logically match what was happening before.
